WalletWasabi/Nito/AsyncEx/AsyncMutex.cs

```diff
@@ -0,0 +1,286 @@
+
+		public AsyncMutex(string name)
+		{
+			// https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex?view=netframework-4.8
+			// On a server that is running Terminal Services, a named system mutex can have two levels of visibility.
+			// If its name begins with the prefix "Global\", the mutex is visible in all terminal server sessions.
+			// If its name begins with the prefix "Local\", the mutex is visible only in the terminal server session where it was created.
+			// In that case, a separate mutex with the same name can exist in each of the other terminal server sessions on the server.
+			// If you do not specify a prefix when you create a named mutex, it takes the prefix "Local\".
+			// Within a terminal server session, two mutexes whose names differ only by their prefixes are separate mutexes,
+			// and both are visible to all processes in the terminal server session.
+			// That is, the prefix names "Global\" and "Local\" describe the scope of the mutex name relative to terminal server sessions, not relative to processes.
+			ShortName = name;
+			FullName = $"Global\\4AA0E5A2-A94F-4B92-B962-F2BBC7A68323-{name}";
+			Mutex = null;
+			MutexThread = null;
+
+			// If we already have an asynclock with this fullname then just use it and don't create a new one.
+			lock (AsyncMutexesLock)
+			{
+				if (AsyncMutexes.TryGetValue(FullName, out AsyncMutex asyncMutex))
+				{
+					AsyncLock = asyncMutex.AsyncLock;
+				}
+				else
+				{
+					AsyncLock = new AsyncLock();
+
+					AsyncMutexes.Add(FullName, this);
+				}
+			}
+		}
+
+		private int _command;
+		private ManualResetEvent ToDo { get; } = new ManualResetEvent(false);
+		private ManualResetEvent Done { get; } = new ManualResetEvent(false);
+
+		private Exception LatestHoldLockException { get; set; }
+		private object LatestHoldLockExceptionLock { get; } = new object();
+
+		private void HoldLock(object cancellationTokenObj)
+		{
+			CancellationToken ct = cancellationTokenObj is CancellationToken ?
+				(CancellationToken)cancellationTokenObj :
+				CancellationToken.None;
+
+			while (true)
+			{
+				try
+				{
+					ToDo.WaitOne();
+
+					if (Interlocked.CompareExchange(ref _command, 0, 1) == 1)
+					{
+						Mutex = new Mutex(initiallyOwned: true, FullName, out bool createdNew);
+						if (createdNew)
+						{
+							continue;
+						}
+						else
+						{
+							DateTime start = DateTime.Now;
+							bool acquired = false;
+
+							while (DateTime.Now - start > TimeSpan.FromSeconds(90))
+							{
+								acquired = Mutex.WaitOne(1000);
+
+								if (acquired)
+								{
+									break;
+								}
+
+								if (ct != CancellationToken.None)
+								{
+									if (ct.IsCancellationRequested)
+									{
+										throw new OperationCanceledException();
+									}
+								}
+							}
+
+							if (acquired)
+							{
+								continue;
+							}
+						}
+					}
+					else if (Interlocked.CompareExchange(ref _command, 0, 2) == 2)
```

Comment:
TBH I don't understand this class, in part because I fail to understand what these `_command` numbers are. It would help to use an enumeration, constant or at least to add a few comments.
>Always code as if the guy who ends up maintaining your code will be a violent psychopath who knows where you live.

Reply:
The main point here is that you have to release the mutex on the same thread you acquired it on. I myself am familiarizing with this logic, too, I'm not very happy about it either, David wrote it. However this is the only correctly working result of almost a week of going back and forth from Mutex, Semaphore, WaitHandle and many  synchronization primitives and concepts. This is the only approach that's working properly on all platforms.

Reply:
Btw, ACK for more comments. @molnard can you go ahead and do it?

Reply:
Done!